#181 - move ktfmt to use Worker API for classloader isolation

cloudshiftchris opened the pull request:
<!-- Thanks for taking the time to write this Pull Request ❤️ -->

## 🚀 Description
This PR moves `ktfmt` to use the Worker API for class loader isolation to mitigate class path collisions with differing Kotlin/Gradle versions.

## 📄 Motivation and Context
Addresses #181 and possibly #157, and likely other latent issues cause by class path conflicts.

## 🧪 How Has This Been Tested?
All unit and integration tests updated to reflect this change.

## 📦 Types of changes
- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

cloudshiftchris commented:
Notes on this comment:
>  I've been investigating isolating the KtFmt dependencies so that users could effectively specify the tool versions without having a 1:1 relationship between gradle plugin-ktfmt tool.

While this PR addresses running ktfmt in an isolated class loader, it doesn't directly provide the capability to alter the ktfmt version separate from the plugin, though it does lay the groundwork for it.

To do so requires addressing two challenges:

1. Binary compatibility with currently supported and future ktfmt versions;
2. Semantic compatibility with currently supported and future ktfmt versions;

Binary compatibility can be handled by providing `adapter` interfaces for versions of ktfmt that alter the public API (it's a small footprint, but does include things like constants, enumerations, etc).  This can be seen, for example, in Spotless where ktlint / ktfmt versions are specified in a Gradle extensions thru to selecting an appropriate adapter, etc.

Semantic compatibility, in this context, refers to maintaining expected behaviour for current and future ktfmt versions.  This is accomplished by having a robust test suite (large set of data to format, across different format styles / configurations, across different ktfmt versions), with deviations addressed by either the aforementioned adapters or 'fyi' documentation/release note updates.

cortinico commented on plugin-build/plugin/src/main/java/com/ncorti/ktfmt/gradle/tasks/KtfmtCheckTask.kt:
```diff
@@ -1,52 +1,43 @@
+        val results = inputFiles.submitToQueue(workQueue, KtfmtCheckAction::class.java)
 
-        val failures = result.filterIsInstance<KtfmtFailure>()
+        logger.d("Format results: $results")
```
```suggestion
        logger.d("Check results: $results")
```

cortinico commented on plugin-build/plugin/src/main/java/com/ncorti/ktfmt/gradle/tasks/KtfmtWorkAction.kt:
```diff
@@ -0,0 +1,159 @@
+ * transitive dependencies)
+ */
+internal abstract class KtfmtWorkAction : WorkAction<KtfmtWorkAction.KtfmtWorkParameters> {
+    protected val logger = Logging.getLogger(KtfmtWorkAction::class.java)
```
```suggestion
    protected val logger: Logger = Logging.getLogger(KtfmtWorkAction::class.java)
```

cortinico commented on plugin-build/plugin/src/main/java/com/ncorti/ktfmt/gradle/tasks/KtfmtWorkAction.kt:
```diff
@@ -0,0 +1,159 @@
+        return result
+    }
+
+    private fun writeResult(result: KtfmtResult) {
+        val resultFile = parameters.workingDir.asFile.get().resolve(UUID.randomUUID().toString())
+        resultFile.writeText(result.toResultString())
+    }
```
Not a super big fan of having to write at most 2x file for each file we reformat. Could we just create a `ktfmt-results` file inside the `workingDir` and append there the results?

cloudshiftchris replied:
this won't be a material impact on performance.  Kept this down to a simple CSV (three fields) to avoid overhead from say Json etc.

We can't append to existing file as each format is handled in a Worker and may be run in parallel, in different threads.

cortinico replied:
> We can't append to existing file as each format is handled in a Worker and may be run in parallel, in different threads.

Yup you're right. Let's keep it as it is then

cortinico commented on plugin-build/plugin/src/main/java/com/ncorti/ktfmt/gradle/tasks/KtfmtWorkAction.kt:
```diff
@@ -0,0 +1,159 @@
+    }
+}
+
+// used for marshalling results from WorkAction back to caller
```
Could we have those classes in separate files within the same package (maybe we can have a `.worker` package)?

cloudshiftchris replied:
Good idea; the work actions and their call chain (KtfmtFormatter) are now in `.worker` package.  This should be the only package that imports, uses ktfmt directly to maintain isolation.

cortinico reviewed (approved):
Amazing work @cloudshiftchris! Thank you very much for doing this 👍 really appreciate your time.

I've left several comments, which are mostly questions. Once this is all sorted out we can merge it and I can make a new release just after.

cloudshiftchris commented on gradle/libs.versions.toml:
```diff
@@ -1,7 +1,6 @@
 [versions]
-agp = "8.1.0"
+agp = "8.0.2"
```
Yes, once everything else is addressed will revert that (otherwise I can't use IntelliJ!)

cloudshiftchris commented on plugin-build/plugin/src/main/java/com/ncorti/ktfmt/gradle/KtfmtPlugin.kt:
```diff
@@ -24,26 +26,50 @@ abstract class KtfmtPlugin : Plugin<Project> {
-                project.logger.i("Skipping Android task creation, as KMP is applied")
-            } else {
-                applyKtfmtToAndroidProject(project, ktfmtExtension, topLevelFormat, topLevelCheck)
+            project.dependencies.add(ktFmt.name, "com.facebook:ktfmt:0.44")
```
should be able to put this in version catalog to align with everything else, so there is only a single ktfmt version.  Good catch.

cloudshiftchris replied:
This is now pulled from the singular `ktfmt` version catalog entry, albeit awkwardly due to something not right with the code base (see comment; will open a separate issue and address that once this PR is complete).

cloudshiftchris replied:
This is now consistent with the version catalog entry as the single source of truth for the ktfmt version.

cortinico replied:
Awesome thank you!

cloudshiftchris commented on plugin-build/plugin/src/main/java/com/ncorti/ktfmt/gradle/tasks/KtfmtBaseTask.kt:
```diff
@@ -1,40 +1,37 @@
+        includeOnly.convention("")
+    }
+
+    @get:InputFiles internal abstract val ktfmtClasspath: ConfigurableFileCollection
```
Not sure, will research.  The Gradle docs for using the Worker API show it this way.

cloudshiftchris replied:
Ok: `@get:Classpath` normalizes the entries (ignoring, for example, timestamp differences in manifest entries) for caching purposes.  It is complementary to `@get:InputFiles` which marks the files as input.
Have added them both.

cloudshiftchris commented on plugin-build/plugin/src/main/java/com/ncorti/ktfmt/gradle/tasks/KtfmtBaseTask.kt:
```diff
@@ -55,66 +51,27 @@ abstract class KtfmtBaseTask : SourceTask() {
-        // before falling back to default values.
+    private fun formattingOptions(): FormattingOptionsBean {
         if (bean == null) {
+            // TODO - set this as property / convention
```
I will dig into this.

cloudshiftchris replied:
This is a little bigger than would prefer to tackle right now; effectively it involves annotating (`@get:Input`) the extension properties and passing the extension into the task, rather than the `FormattingOptionsBean` object.  Once into the execution phase we can convert into the `FormattingOptionsBean` and continue from there, preserving downstream code.
Will open a separate issue for this.

cortinico replied:
Should we open this one?

cloudshiftchris commented:
@cortinico have addressed all PR feedback; will create a couple issues for later cleanup:

- source files are in `java` folder which is preventing use of Gradle Kotlin DSL, resulting in more verbose code;
- revisit `FormattingOptionsBean` and how that is passed to tasks

Should be good-to-go for isolated class loader changes.

cortinico commented:
> will create a couple issues for later cleanup:

@cloudshiftchris do you need a new release with just this change or are you planning to follow-up on those items so we can make a release after?
